python/equistore-operations/equistore/operations/join.py

```diff
@@ -285,3 +303,118 @@ def join(
         return remove_dimension(tensor_joined, name="tensor", axis=axis)
     else:
         return tensor_joined
+
+
+def _tensors_intersection(tensors: List[TensorMap]) -> List[TensorMap]:
+    """Create a new tensors list where keys are based on the intersection from all
+    tensors.
+
+    Blocks corresponding to keys that are not present in all tensor will be discarded.
+    """
+    # Construct a Labels object with intersected keys
+    all_keys = tensors[0].keys
+    for tensor in tensors[1:]:
+        all_keys = all_keys.intersection(tensor.keys)
+
+    # Create new blocks and discard bocks not present in all_keys
+    new_tensors = []
+    for tensor in tensors:
+        new_blocks = [tensor.block(key).copy() for key in all_keys]
+        new_tensors.append(TensorMap(keys=all_keys, blocks=new_blocks))
+
+    return new_tensors
+
+
+def _tensors_union(tensors: List[TensorMap], axis: str) -> List[TensorMap]:
+    """Create a new tensors list where keys are based on the union from all tensors.
+
+    Missing keys will be filled by empty blocks having containing no labels in the
+    ``axis`` dimension.
+    """
+    # Constrauct a Labels object with all keys
+    all_keys = tensors[0].keys
+    for tensor in tensors[1:]:
+        all_keys = all_keys.union(tensor.keys)
+
+    # Create empty blocks for missing keys for each TensorMap
+    new_tensors = []
+    for tensor in tensors:
+        _, map, _ = all_keys.intersection_and_mapping(tensor.keys)
+
+        missing_keys = Labels(
+            names=tensor.keys.names, values=all_keys.values[map == -1]
+        )
+
+        new_keys = tensor.keys.union(missing_keys)
+        new_blocks = [block_to(block) for block in tensor]
```

Comment:
why are we doing `block_to` here? Especially without any argument, that will not do much? If we need a copy of the block, we can call `block.clone()`